Earpiece switcher and overlay

robintown opened the pull request:
A couple things still need doing here:
- [x] Write tests
  - move mediaDevice test to callViewModel
- [x] Remove the controls API for toggling the earpiece switcher, which we no longer need
- [x] ~~Split the header out into another PR, maybe?~~
- [x] Modernize the new context
- [x] Force dark theme for the overlay (and also the header and footer, when it is active)
   - The light theme looks very correct to me when the phone is in light theme mode (this is tbd)
   - Let's ask Aaron
   - Result: We do the same then what we do with emoji and settings overlays: dark overlay in light and dark theme
     This requires us to also swap the icon colors for the header bar icons
   - Result V2: we are fine with the inconsistency between modal background darkening and earpiece overlay lightening
- [x] Fix android designs (video goes to top of the bar)
- [x] Fix hover border being clipped
- [x] Fix header being shown in PiP
- [x] Only present the fast toggling UX in case either the Earpiece speaker or the Main speaker are selected
    - The fast toggling UX is only toggling between Earpiece and Main speaker
    - We also consider always showing the toggles but making them actions (shows what will happen when pressed instead of the current state) This is currently implemented in the pr as of: https://github.com/element-hq/element-call/pull/3347/commits/df3f962f4707bccc6100001ad1b9bffee95d5d0c
    - This might be not enough: we also need to make sure to always show the button (even if in bluetooth mode)
- [x] landscape should hide the header for now
- [x] Update Element Web version used for end-to-end tests to a version that includes https://github.com/element-hq/element-web/pull/30213

Potential candidates for a follow up PR
    
- [x] fix android designs OneOnOneLayout (so that the app_bar overlays the video and but the floating tile does still respect the bounds)
- [x] fix OneOnOneLayout Landscape (currently the app bar is not visible in this mode.

Closes https://github.com/element-hq/element-call/issues/3271

robintown committed:
Tidy up controls.ts

robintown committed:
Add a global control for toggling earpiece mode

This will be used by Element X to show an earpiece toggle button in the header.

robintown committed:
Add an earpiece overlay

toger5 commented on docs/controls.md:
```diff
@@ -23,3 +23,4 @@ On mobile platforms (iOS, Android), web views do not reliably support selecting
   playing audio in the webview. It can be helpful to do device setup on the native app when the webviews audio is ready.
   In particular android is using it to setup the output channel so that the call volume can
   be controlled by the hardware volume rocker.
+- `controls.toggleEarpieceMode(): void` Switches audio between the earpiece device and an arbitrary non-earpiece device.
```
Do we really need this. It is for the native OS to enforce EC to switch to earpiece mode.
Do we have any UX branch that requires this change to get initiated by the native os?
On android we can just do so by setting the device right?
On iOS the only scenario I can imagine is using the proximity sensor but that is not active if we are not yet in earpiece mode.

robintown replied:
We don't need this, it is left over from when I thought the button was going to have to be in the native header

robintown committed:
WIP: Add app bar and earpiece toggle button

robintown commented:
@fkwp This is now ready for testing! Expect to see two headers on mobile until we get to the EX tasks listed in https://github.com/element-hq/voip-internal/issues/360

toger5 committed:
Fix header
The header needs to be passed forward as a string to some components and as a bool (hideHeader) to others.
Also use a enum instead of string options.

toger5 committed:
fix top clipping with header

toger5 committed:
fix styling
 - make video render below the header on Android.

toger5 committed:
hide app bar in pip

toger5 committed:
revert android overlay app_bar

toger5 committed:
Modernize AppBarContext

toger5 committed:
Style header icon color as desired and switch earpice/speaker icon

toger5 committed:
fix initial selection when using controlled media

toger5 committed:
Add "Back to video" button

toger5 committed:
fix lints

toger5 committed:
fix tests

toger5 committed:
remove dead code

toger5 committed:
prioritize header > hideHeader

toger5 committed:
add snapshot test

toger5 committed:
fix back to video button

toger5 committed:
fix logic for earpiece quick switcher to be based on switching target
(also show quick switcher if headphones are connected)

robintown committed:
Merge branch 'livekit' into robin/earpiece-overlay

robintown committed:
Request capability to learn the room name

We now need the room name in order to implement the mobile (widget-based) designs with the app bar.

robintown committed:
Test the CallViewModel output switcher directly
